cicerone/references/reference_resolver.py

```diff
@@ -0,0 +1,344 @@
+            The $ref string if found, None otherwise
+        """
+        return getattr(model, "__pydantic_extra__", {}).get("$ref")
+
+    def _try_resolve_ref(self, ref: str) -> typing.Any | None:
+        """Attempt to resolve a reference, returning None on failure.
+
+        Args:
+            ref: Reference string to resolve
+
+        Returns:
+            Resolved object or None if resolution fails
+        """
+        try:
+            return self.resolve_reference(ref, follow_nested=True)
+        except (ValueError, RecursionError):
+            return None
+
+    def _resolve_model_field(
+        self, parent_obj: pydantic.BaseModel, field_name: str, field_value: pydantic.BaseModel
+    ) -> None:
+        """Resolve a Pydantic model field that may contain a $ref.
+
+        Args:
+            parent_obj: The parent object containing this field
+            field_name: Name of the field being resolved
+            field_value: The Pydantic model value to check for $ref
+        """
+        if ref := self._get_ref_from_model(field_value):
+            if resolved := self._try_resolve_ref(ref):
+                setattr(parent_obj, field_name, resolved)
+        else:
+            setattr(parent_obj, field_name, self._resolve_nested_references(field_value))
+
+    def _resolve_container(self, container: dict | list) -> None:
+        """Resolve references in dictionary or list containers.
+
+        Args:
+            container: Dictionary or list to process
+        """
+        items = container.items() if isinstance(container, dict) else enumerate(container)
+
+        for key, value in items:
+            if isinstance(value, pydantic.BaseModel):
+                if ref := self._get_ref_from_model(value):
+                    if resolved := self._try_resolve_ref(ref):
+                        container[key] = resolved
+                else:
+                    container[key] = self._resolve_nested_references(value)
+
+    def get_all_references(
+        self,
+        obj: typing.Any | None = None,
+        visited: set[int] | None = None,
+    ) -> dict[str, spec_reference.Reference]:
+        """Find all references in an object or the entire spec.
+
+        Recursively searches for all $ref keywords in the given object or the entire spec.
+
+        Args:
+            obj: Object to search for references (defaults to entire spec)
+            visited: Set of object ids already visited (for circular reference handling)
+
+        Returns:
+            Dictionary mapping $ref strings to Reference objects
+
+        Example:
+            >>> resolver = ReferenceResolver(spec)
+            >>> all_refs = resolver.get_all_references()
+            >>> # Access by reference string
+            >>> user_ref = all_refs.get('#/components/schemas/User')
+            >>> # Get all local references
+            >>> local_refs = {k: v for k, v in all_refs.items() if v.is_local}
+        """
+        obj = obj or self.spec.raw
```

Comment:
Again - if we are able to build some sort of property that has the references instead of trawling through the raw property it feels like we could do more advantageous stuff. Thoughts?